### ISSUE TITLE
Share Notebooks Feature Request

### ISSUE THREAD

**Issue by alanedelman** [JuliaLang/JuliaBox#217]
students want to show me what they are doing
too much friction right now
already google hangout screen sharing is handy for this
but requires synchrony and doesn't let me add my two cents

doesn't have to be as good as google docs --- for example,
people could live with requiring a refresh if that makes it much easier

**Comment by shashi** [comment]
@tanmaykm How can we take a specific file out of a container and place it under a publicly accessible URL? Would it be OK that when someone shares a notebook, a copy is made to a public location? The URL could include a hash of the contents of the notebook, this way every time someone saves a new version of a notebook, they can get a completely new share URL for that version. a notebook URL once generated will point to the same content and hence there won't be confusion due to the time at which a student opened the shared link.

**Comment by tkelman** [comment]
There's a really handy ruby gem where you can just run `gist nb.ipynb`, it uploads to a gist and gives you back a URL.

**Comment by lilinjn** [comment]
Hi -- I find myself wanting a similar feature -- but perhaps going in the other direction....

Concretely, I'm looking for a feature or set of features that emulate an important part of the user experience offered by http://nbviewer.ipython.org/ -- which is to be able to easily allow others to visualize a notebook  without having them to know much about julia or even git.  Key to this flow is the ability to create restful url's that nbviewer knows how to intelligently parse to retrieve the document's source.

In my case, many of my collaborators work in scheme or racket, and I'm the odd one out using Julia -- I've found it extremely useful being able to send them urls which they can click on and see a fully rendered notebook via nbviewer (not surprisingly, this has also been a non-preachy way of exposing the potential benefits of Julia to folk who might otherwise have not been interested) 
 
However, I find myself using Interact (great work, sashi!) more and more, and so nbviewer non-longer cuts it -- what would be great is to be able to use juliabox as an alternate.  

Concretely, I think the user experience needs to be as simple as possible
++ click a link
++ enter google account info (this is totally fine)
++ see the running notebook.  The notebook could either be located at a url for a github blob  or a gist, or possibly a shared notebook from a specific juliabox user's persistent filesystem.

++ from there, it seems like you could add support for cloning a notebook so that people could make changes that are easily communicated back to the owner of the original notebook (One could either use a google drive/dropbox model for simplicity or go the github fork/PR route )

I think this user experience deserves a name, and I would suggest "ad-hoc" as opposed to the current experience which is "persistent" e.g. juliabox keeps a file system for you across sessions, etc.  The "persistent" experience seems great for supporting a semester long class, but "ad-hoc" might be a great way to help spread the word about julia.  There is no reason that ad-hoc and persistent have to be incompatible or not well integrated (e.g. think gist and github repo), but I think giving them separate names helps clarify the purpose -- "Ad-hoc" exists to let casual users quickly, and with minimal fuss, work with notebooks that use features such as Interact.jl )

I'm willing to pitch in if there is interest is going down this path.  

Thanks for all the great work!  cheers

**Comment by tkelman** [comment]
Is there / could there be an "import notebook from url" feature?

**Comment by tanmaykm** [comment]
I think simple public file sharing can be done with:
- have a public S3 bucket where we place files that are shared from within JuliaBox and give out its HTTP URL
- have ability to fetch a notebook from a URL and run it.

This is somewhat possible today with github/gdrive/pastebin and `wget`/`curl` (though on the console only). We should probably document these somewhere.

Limited sharing (with designated persons / password) will address the original requirement of @alanedelman. The ux as described by @lilinjn is important too. This will require us to maintain records of files shared and handle download requests.

**Comment by ViralBShah** [comment]
I think an "import from url" is a good simple feature to have. With Google Drive / S3 / etc., people can easily share URLs of their files with others. So, making it easy to pull a notebook / files in through a URL would be a great start.

**Comment by lilinjn** [comment]
I think 'import from url' seems like a great starting point.

**Comment by lgekeler** [comment]
This project has tremendous merit for teaching. I'm an LMS Admin looking for a way for STEM profs who know python to, not just teach python, per se, but write their course materials (think textbook) in python NBs. 
This doesn't create yet another learning management system, we have those, instead I'm describing authorware for the course. Text, code, graphs, variables you can manipulate, graphs redraw, etc - all in demonstration mode. Above I believe this is what you've named a "persistent URL."
Now in teaching eventually the Instructor writes assignments and expect submissions back. The student forks the assignment, writes their bit, and then... how would you suggest they submit? Again, I think its a persistent URL, the student submits the URL to their learning management system's assignment tool. This is where it gets fuzzy for me. Several requirements are needed, some already in the learning management system. Here's a tentative workflow
1) Instructor or TA clicks on submitted URL in the learning management system and is taken to a read-only persistent copy, the assignment submission.
2) In another browser, inside the learning management system, they write their feedback, maybe they take screen captures of the assignment submission screen and markup their screen cap? Or, do they again make a copy of the student assignment submission which they could write to and do their feedback there? The Instructor only puts the URL to the feedback copy in the feedback portion of the learning management system.
3) Instructor assigns a grade in the learning management system.

I'm looking for an elegant solution to #2. Maybe the integration with Google Docs supplies it?

**Comment by tkelman** [comment]
@lgekeler you may want to take a look at https://github.com/jupyter/nbgrader.

**Comment by lgekeler** [comment]
Thank you. Has anyone used it with juliabox?

**Comment by tanmaykm** [comment]
https://github.com/shashi/Homework.jl by @shashi does something similar on JuliaBox.

**Comment by tkelman** [comment]
I don't think so. The code in nbgrader may be a bit too Python-specific to make sense here. Homework.jl is the closest Julia-specific thing but looks like it could use some documentation.

**Comment by lgekeler** [comment]
Thanks. Google auth and Google Drive integration are missing from the JupyterHub projects I'm aware of, but seriously important for google schools to implement. You're doing great work.

**Comment by ViralBShah** [comment]
Homework.jl has been used at least in two of @alanedelman 's classes at MIT. It works well, but it is a bit of a hack. @shashi can say more.

**Comment by naartjie** [comment]
Just came across this issue now. Is sharing still being considered? 

Out of interest, are there any alternatives to JuliaBox? I found https://colab.research.google.com/ today, it looks good, but doesn't support Julia.

**Comment by djsegal** [comment]
How do people share interactive jupyter notebooks now?

// this puts a lot of strain on relations with other scientists that don't use Julia yet
